streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -2472,18 +2472,58 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
     }
 
   /**
-   * Interrupts the stream if it does not produce a value after d duration.
+   * Ends the stream if it does not produce a value after d duration.
    */
   final def timeout(d: Duration): ZStream[R with Clock, E, O] =
     ZStream[R with Clock, E, O] {
       self.process.map { next =>
         next.timeout(d).flatMap {
-          case Some(a) => ZIO.succeedNow(a)
-          case None    => ZIO.interrupt
+          case Some(a) => Pull.emit(a)
+          case None    => Pull.end
         }
       }
     }
 
+  /**
+   * Ends the stream with given error if it does not produce a value after d duration.
+   */
+  final def timeoutErr[E1 >: E](e: E1)(d: Duration): ZStream[R with Clock, E1, O] =
+    ZStream[R with Clock, E1, O] {
+      self.process.map { next =>
+        next.timeout(d).flatMap {
+          case Some(a) => Pull.emit(a)
+          case None    => Pull.fail(e)
+        }
+      }
+    }
+
+  /**
+   * Switches the stream if it does not produce a value after d duration.
+   */
+  final def timeoutTo[R1 <: R, E1 >: E, O2 >: O](
+    d: Duration
+  )(that: ZStream[R1, E1, O2]): ZStream[R1 with Clock, E1, O2] =
+    ZStream[R1 with Clock, E1, O2] {
+      for {
+        currStream   <- Ref.make[ZIO[R1, Option[E1], Chunk[O2]]](Pull.end).toManaged_
```

Comment:
Is it possible to use catchAllCause for this combinator? Implementation looks fairly similar.

Reply:
I will take a look

Reply:
Given that `timeoutErrorCause` has been implemented, the following implementation comes to mind:

```scala
final def timeoutTo[R1 <: R, E1 >: E, O2 >: O](
    d: Duration
)(that: ZStream[R1, E1, O2]): ZStream[R1 with Clock, E1, O2] = {
  object StreamTimeout extends Throwable
  self.timeoutErrorCause(Cause.die(StreamTimeout))(d).catchSomeCause { case Cause.Die(StreamTimeout) => that }
}
```

It requires implementing `catchSomeCause` though which I can add (maybe in a separate PR). @iravid what do you think?

Reply:
Sounds good for a follow up if you have the cycles later on. The current implementation looks good for this PR! Thanks!

Reply:
Cool, I will submit a separate PR later with `catchSomeCause` and the refactor of `timeoutTo`